sql/logic_test.go

```diff
@@ -0,0 +1,318 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Peter Mattis (peter@cockroachlabs.com)
+
+package sql_test
+
+import (
+	"bufio"
+	"bytes"
+	"crypto/md5"
+	"database/sql"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"reflect"
+	"regexp"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/cockroachdb/cockroach/server"
+	"github.com/cockroachdb/cockroach/testutils"
+	"github.com/cockroachdb/cockroach/util/leaktest"
+)
+
+var (
+	resultsRE = regexp.MustCompile(`^(\d+)\s+values?\s+hashing\s+to\s+([0-9A-Fa-f]+)$`)
+	errorRE   = regexp.MustCompile(`^(?:statement|query)\s+error\s+(.*)$`)
+)
+
+type lineScanner struct {
+	*bufio.Scanner
+	line int
+}
+
+func newLineScanner(r io.Reader) *lineScanner {
+	return &lineScanner{
+		Scanner: bufio.NewScanner(r),
+		line:    0,
+	}
+}
+
+func (l *lineScanner) Scan() bool {
+	ok := l.Scanner.Scan()
+	if ok {
+		l.line++
+	}
+	return ok
+}
+
+type logicStatement struct {
+	pos       string
+	sql       string
+	expectErr string
+}
+
+type logicQuery struct {
+	pos             string
+	sql             string
+	expectErr       string
+	expectedValues  int
+	expectedHash    string
+	expectedResults []string
+}
+
+type logicValue string
+
+func (v *logicValue) Scan(value interface{}) error {
```

Comment:
Yeah, that's where the idea for this was gleaned from. Might be worth moving this type into the `sql` package so that it can be used by sql cli as well. I'll add a TODO.